### PR TITLE
sync_channel

### PR DESCRIPTION
**Pull request by kteza1**
Hi. I don't know much about the architecture of mioco but I tried to replicate `Sender` code to add `SyncSender`. I'm not sure about the implications but example seems to be working on linux.

### CODE REVIEW

File: src/sync/mpsc.rs [hunk 1]
```diff
@@ -188,6 +193,33 @@ impl<T> Sender<T> {
...
+            }),
+            sender: sender,
+        }
+    }
+
+    /// Deliver `T` to the other end of the channel.
+    ///
+    /// Channel behaves like a queue.
+    ///
+    /// This is non-blocking operation.
```
Comment (dpc) [inline src/sync/mpsc.rs]: As you're using native `SyncSender`, below `send` can now block.

File: src/sync/mpsc.rs [hunk 2]
```diff
@@ -188,6 +193,33 @@ impl<T> Sender<T> {
...
+    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
+        try!(self.sender.send(t));
```
Comment (dpc) [inline src/sync/mpsc.rs]: This is blocking now.

**Comment by dpc** [comment]
This implementation will work only if sender is used outside of mioco. If `SyncSender` is used inside of mioco it must mever block on `send()`.

The most naive (but acceptable for now) solution would be to use:

https://doc.rust-lang.org/std/sync/mpsc/struct.SyncSender.html#method.try_send

if running inside mioco (can be detected with http://dpc.pw/mioco/mioco/fn.in_coroutine.html), use `try_send` and if channel was full, just http://dpc.pw/mioco/mioco/fn.yield_now.html to retry later. It's not optimal but it will work well enough and technically correct.

Also, the current policy is that there must be some unit-test for every feature. In this case:

* one test if this simply works
* one test  if it handles full-channel condition correctly

After that I'd be happy to merge. Thanks for your contribution.

**Comment by kteza1** [comment]
Got it. Thanks for the review. I'll make the changes you suggested :)

File: tests.rs [hunk 3]
```diff
@@ -1272,6 +1299,49 @@ fn mpsc_inside_inside() {
...
+        let finished_copy1 = finished_ok1.clone();
+
+        let finished_ok2 = Arc::new(Mutex::new(false));
+        let finished_copy2 = finished_ok2.clone();
+
+        mioco::start_threads(threads, move || {
+            mioco::spawn(move || {
+                for i in 0..10 {
+                    println!("#### Sending ...");
+                    tx1.send(i).unwrap();
```
Comment (kteza1) [inline tests.rs]: There seems to be some kind of problem here. I see an yield happening here but this coroutine isn't getting invoked again.

But yield is working properly [here](https://github.com/dpc/mioco/pull/148/commits/2d81d5ff2b5f9db4efcb2648c804ea86645b81d7#diff-80f8aab3e0907ed73ef080518e07da7dR1255)

Comment (kteza1) [inline tests.rs]: Maybe recv() is blocking.

File: src/sync/mpsc.rs [hunk 4]
```diff
@@ -208,9 +218,26 @@ impl<T> SyncSender<T> {
...
     /// Channel behaves like a queue.
     ///
-    /// This is non-blocking operation.
-    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>> {
-        try!(self.sender.send(t));
+    /// When sender is outside mioco, this is a blocking operation (if channel
+    /// is full). If the sender is inside mioco, `try_send` will be
+    /// called where coroutine is yielded incase of `TrySendError::Full` error.
+    pub fn send(&self, t: T) -> Result<(), mpsc::SendError<T>>
+    where T: Clone {
```
Comment (dpc) [inline src/sync/mpsc.rs]: You should not need the `Clone`

File: src/sync/mpsc.rs [hunk 5]
```diff
@@ -208,9 +218,26 @@ impl<T> SyncSender<T> {
...
+       if in_coroutine() {
+            loop {
+                match self.sender.try_send(t.clone()) {
```
Comment (dpc) [inline src/sync/mpsc.rs]: `TrySendError` returns the original value, just change `t: T` to `mut t: T`, and assign the returned value back to `t`.

File: tests.rs (hunk 3 above)
Comment (dpc) [inline tests.rs]: Will investigate today evening.

**Comment by dpc** [comment]
Nice. Tests and everything. We'll help with investigating the inside-inside issue, and then we can merge.

File: tests.rs [hunk 6]
```diff
@@ -1216,6 +1216,20 @@ fn mpsc_outside_outside() {
...
+#[test]
+fn mpsc_sync_outside_outside() {
+    let (tx1, rx1) = mpsc::sync_channel(3);
+    thread::spawn(move || {
+        for i in 0..10 {
+            tx1.send(i).unwrap();
+        }
+    });
+    for i in 0..10 {
+        thread::sleep(Duration::new(1, 0));
```
Comment (Ryman) [inline tests.rs]: Could these sleeps be taken out of the loops in these tests or are they exercising something specific?

Comment (kteza1) [inline tests.rs]: @Ryman Just wanted to ensure that tx speed is > rx. Wouldn't make much diff for testcases but if someone wants to debug more, adding prints at `send` and `recv` shows that sends are being blocked.

**Comment by dpc** [comment]
Using `writeln!(&mut std::io::stderr(), "{}. #### Sending ...", i);` in tests instead of println! is much better way, as println is buffered.

I think I've found the issue, BTW.

File: src/sync/mpsc.rs [hunk 7]
```diff
@@ -188,6 +189,73 @@ impl<T> Sender<T> {
...
+       if in_coroutine() {
+            loop {
+                t = match self.sender.try_send(t) {
+                    Err(mpsc::TrySendError::Full(t)) => {
+                        println!("Yielding ...");
+                        yield_now();
+                        t
+                    }
+                    Err(mpsc::TrySendError::Disconnected(t)) => return Err(mpsc::SendError(t)),
+                    Ok(t) => return Ok(t),
```
Comment (dpc) [inline src/sync/mpsc.rs]: This should be `break`. Otherwise the below code is not exectued.

File: src/sync/mpsc.rs [hunk 8]
```diff
@@ -188,6 +189,73 @@ impl<T> Sender<T> {
...
+                };
+            }
+        } else {
+            try!(self.sender.send(t));
+        }
+
+        let prev_counter = self.shared.counter.fetch_add(1, Ordering::SeqCst);
```
Comment (dpc) [inline src/sync/mpsc.rs]: This code. Receiver will never get notified to wake up.

**Comment by dpc** [comment]
I've posted comment. After fixing that, all tests pass. Sorry for going back to you after a longer time. I'm just extremely occupied with things.

Anyway, please fix, cleanup, check again, and I'll be super happy to merge! Looking forward to merge more.

**Comment by dpc** [comment]
I added a cleanup patch and pushed to master. Thanks!

**Comment by kteza1** [comment]
Sorry. I didn't get time to look into this. Thanks for taking it up :)

**Pull request closed without merging**